Fix builder function return type

nikeee opened the pull request:
The return value of the builder function seems to be wrong.
The docs state:
> `builder` can also be a function. This function is executed with a yargs instance, and can be used to provide advanced command specific help:

This code sample does not compile because the return type of the builder function does not match:
```TypeScript
yargs
  .command(
    'get',
    'make a get HTTP request',
    function (yargs) {
      return yargs.option('u', {
        alias: 'url',
        describe: 'the URL to make an HTTP request to'
      })
    },
    function (argv) {
      console.dir(argv)
    }
  )
  .help()
  .argv
```
It should be an instance of `Argv`, not `Options`. This PR fixes this.

- [x] Prefer to make your PR against the `types-2.0` branch.
- [x] Test the change in your own code.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [x] Run `npm run lint -- package-name` if a `tslint.json` is present.

If changing an existing definition:
- [x] Provide a URL to  documentation or source code which provides context for the suggested changes: https://www.npmjs.com/package/yargs#commandmodule
- [x] Increase the version number in the header if appropriate.

pushplay commented on yargs/yargs-tests.ts:
```diff
@@ -226,6 +226,33 @@ function command() {
 		.argv;
+
+	yargs
+	.command('get', 'make a get HTTP request', function (yargs) {
```
Chained functions should be indented one more than the start of the statement.

pushplay commented on yargs/yargs-tests.ts:
```diff
@@ -226,6 +226,33 @@ function command() {
+	.command('get', 'make a get HTTP request', function (yargs) {
+		return yargs.option('url', {
+		alias: 'u',
+		default: 'http://yargs.js.org/'
```
These two members should also be indented one more.

nikeee commented:
@pushplay done

pushplay commented:
Cheers.

👍

poelstra commented:
Nice catch indeed, thnx!

:+1:

nikeee commented:
I just noticed that there is something else that does not match the docs.

These overloads of the command method:
```TypeScript
command(command: string, description: string, handler: (args: Argv) => void): Argv;
command(command: string, description: string, builder: (args: Argv) => Argv): Argv;
```
They are ambiguous, so I removed the first one since the docs state that if there is a handler, there must be a builder. It would not be possible to distinguish between a handler and a builder.

I noticed this because the argument of the `handler` function seems to be the parsed args, not an `Argv` instance.
This sample does not compile because `argv.url` is part of the parsed args object, not the `Argv` type of the type definitions:
```TypeScript
yargs
    .command(
        'get',
        'make a get HTTP request',
        function (yargs) {
            return yargs.option('u', {
            alias: 'url',
            describe: 'the URL to make an HTTP request to'
            })
        },
        function (argv) {
            console.log(argv.url)
        }
    )
    .help()
    .argv
```
To fix this, I changed the affected command overloads to be generic (see files changed).

pushplay commented:
> since the docs state that if there is a handler, there must be a builder

That's not how I read it initially, but I tested it in code and that is what's going on.  The `Argv` type not being `argv` really confuses me.

```
command<TArgs>(command: string, description: string, builder: (args: Argv) => Argv, handler: (args: TArgs) => void): Argv;
```

I don't think the generic parameter here adds any type safety.  It doesn't tie any two things together so any function can be `handler` unless you also specify the generic type; which you generally wouldn't.

I think it should be `argv: any` (note the name) to match the member at https://github.com/nikeee/DefinitelyTyped/blob/4bc0d4194164bd898695411c2c4893f0c28aa38e/yargs/index.d.ts#L8 as it's the same object.

nikeee commented:
I changed it to `argv: any`.

pushplay commented:
👍
